File: apps/worker/src/circuit-breaker.ts
/**
 * Circuit Breaker for Worker Processors
 *
 * Implements circuit breaker pattern to prevent cascading failures
 * when external endpoints (webhooks, email servers) are unavailable.
 *
 * States:
 * - CLOSED: Normal operation, all requests pass through
 * - OPEN: Failures exceeded threshold, requests are rejected immediately
 * - HALF_OPEN: Testing if service recovered, limited requests allowed
 */

import type { Redis } from 'ioredis';

export enum CircuitState {
  CLOSED = 'closed',
  OPEN = 'open',
  HALF_OPEN = 'half_open',
}

export interface CircuitBreakerConfig {
  /** Unique name for this circuit (e.g., "webhook:endpoint123") */
  name: string;
  /** Number of failures before opening the circuit */
  failureThreshold: number;
  /** Time in milliseconds before transitioning from OPEN to HALF_OPEN */
  resetTimeout: number;
  /** Number of successful requests in HALF_OPEN needed to close circuit */
  successThreshold: number;
  /** Window in milliseconds to track failures (rolling window) */
  rollingWindowMs: number;
}

interface CircuitStats {
  failures: number;
  successes: number;
  lastFailureTime: number;
  state: CircuitState;
  openedAt: number;
}

const DEFAULT_CONFIG: Omit<CircuitBreakerConfig, 'name'> = {
  failureThreshold: 5,
  resetTimeout: 30000,  // 30 seconds
  successThreshold: 2,
  rollingWindowMs: 60000, // 1 minute
};

/**
 * Redis-backed Circuit Breaker for distributed state
 */
export class CircuitBreaker {
  private readonly redis: Redis;
  private readonly config: CircuitBreakerConfig;
  private readonly keyPrefix: string;

  /**
   * FIX-500-112: Circuit breaker key TTL in seconds (24 hours).
   * Without TTL, abandoned circuit breaker keys persist in Redis forever.
   */
  private static readonly KEY_TTL_SECONDS = 86400;

  constructor(redis: Redis, config: Partial<CircuitBreakerConfig> & { name: string }) {
    this.redis = redis;
    this.config = { ...DEFAULT_CONFIG, ...config };
    this.keyPrefix = `circuit:${this.config.name}`;
  }

  /**
   * Check if a request should be allowed through the circuit
   */
  async canExecute(): Promise<boolean> {
    const state = await this.getState();

    switch (state) {
      case CircuitState.CLOSED:
        return true;

      case CircuitState.OPEN: {
        const stats = await this.getStats();
        const timeSinceOpen = Date.now() - stats.openedAt;

        // Transition to HALF_OPEN if reset timeout has elapsed
        if (timeSinceOpen >= this.config.resetTimeout) {
          // FIX-500-111: Only one caller should transition to HALF_OPEN.
          // Use SET NX to ensure only the first concurrent caller wins.
          const acquired = await this.redis.set(
            `${this.keyPrefix}:half_open_lock`,
            '1',
            'EX', 30,  // lock expires after 30s
            'NX'
          );
          if (!acquired) {
            return false; // Another caller already transitioning
          }
          await this.setState(CircuitState.HALF_OPEN);
          return true; // Allow one request through
        }
        return false;
      }

      case CircuitState.HALF_OPEN:
        // Allow limited requests in half-open state
        return true;

      default:
        return true;
    }
  }

  /**
   * Record a successful request
   */
  async recordSuccess(): Promise<void> {
    const state = await this.getState();

    if (state === CircuitState.HALF_OPEN) {
      // FIX-500-110: Set TTL on successes counter so stale successes
      // from hours ago don't incorrectly close the circuit.
      const successes = await this.redis.incr(`${this.keyPrefix}:successes`);
      await this.redis.expire(`${this.keyPrefix}:successes`, 120); // 2 minute TTL

      if (successes >= this.config.successThreshold) {
        // Circuit recovered, close it
        await this.close();
      }
    }

    // FIX-500-113: In CLOSED state, let the rolling window handle failure
    // expiration naturally instead of wiping all failures on a single success.
  }

  /**
   * Record a failed request
   */
  async recordFailure(): Promise<void> {
    const state = await this.getState();
    const now = Date.now();

    // Track failure timestamp for rolling window
    await this.redis.zadd(`${this.keyPrefix}:failure_times`, now, now.toString());

    // Remove old failures outside the rolling window
    const windowStart = now - this.config.rollingWindowMs;
    await this.redis.zremrangebyscore(`${this.keyPrefix}:failure_times`, '-inf', windowStart);

    // Count failures in the rolling window
    const failureCount = await this.redis.zcard(`${this.keyPrefix}:failure_times`);

    // Update last failure time
    // FIX-500-112: Set TTL on last_failure and failure_times
    await this.redis.set(`${this.keyPrefix}:last_failure`, now.toString(), 'EX', CircuitBreaker.KEY_TTL_SECONDS);
    await this.redis.expire(`${this.keyPrefix}:failure_times`, CircuitBreaker.KEY_TTL_SECONDS);

    if (state === CircuitState.HALF_OPEN) {
      // Any failure in half-open state opens the circuit again
      await this.open();
    } else if (state === CircuitState.CLOSED && failureCount >= this.config.failureThreshold) {
      // Threshold exceeded, open the circuit
      await this.open();
    }
  }

  /**
   * Execute a function with circuit breaker protection
   */
  async execute<T>(fn: () => Promise<T>): Promise<{ ok: true; value: T } | { ok: false; error: Error }> {
    if (!await this.canExecute()) {
      return {
        ok: false,
        error: new Error(`Circuit ${this.config.name} is open`),
      };
    }

    try {
      const result = await fn();
      await this.recordSuccess();
      return { ok: true, value: result };
    } catch (error) {
      await this.recordFailure();
      return {
        ok: false,
        error: error instanceof Error ? error : new Error(String(error)),
      };
    }
  }

  /**
   * Get current circuit state
   */
  async getState(): Promise<CircuitState> {
    const state = await this.redis.get(`${this.keyPrefix}:state`);
    return (state as CircuitState) || CircuitState.CLOSED;
  }

  /**
   * Get circuit statistics
   */
  async getStats(): Promise<CircuitStats> {
    const [state, failures, successes, lastFailure, openedAt] = await this.redis.mget(
      `${this.keyPrefix}:state`,
      `${this.keyPrefix}:failures`,
      `${this.keyPrefix}:successes`,
      `${this.keyPrefix}:last_failure`,
      `${this.keyPrefix}:opened_at`
    );

    return {
      state: (state as CircuitState) || CircuitState.CLOSED,
      failures: failures ? parseInt(failures, 10) : 0,
      successes: successes ? parseInt(successes, 10) : 0,
      lastFailureTime: lastFailure ? parseInt(lastFailure, 10) : 0,
      openedAt: openedAt ? parseInt(openedAt, 10) : 0,
    };
  }

  /**
   * Manually open the circuit (for testing or emergency)
   */
  async open(): Promise<void> {
    await this.setState(CircuitState.OPEN);
    // FIX-500-112: Set TTL on opened_at to prevent forever accumulation
    await this.redis.set(`${this.keyPrefix}:opened_at`, Date.now().toString(), 'EX', CircuitBreaker.KEY_TTL_SECONDS);
    await this.redis.del(`${this.keyPrefix}:successes`);
  }

  /**
   * Manually close the circuit
   */
  async close(): Promise<void> {
    await this.setState(CircuitState.CLOSED);
    await this.redis.del(
      `${this.keyPrefix}:failures`,
      `${this.keyPrefix}:successes`,
      `${this.keyPrefix}:failure_times`,
      `${this.keyPrefix}:opened_at`
    );
  }

  /**
   * Reset the circuit (clear all state)
   */
  async reset(): Promise<void> {
    await this.close();
  }

  private async setState(state: CircuitState): Promise<void> {
    // FIX-500-112: Set TTL on state key to prevent forever accumulation
    await this.redis.set(`${this.keyPrefix}:state`, state, 'EX', CircuitBreaker.KEY_TTL_SECONDS);
  }
}

/**
 * Factory for creating circuit breakers with shared Redis instance.
 * Includes LRU eviction to prevent unbounded memory growth.
 */
export class CircuitBreakerFactory {
  private readonly redis: Redis;
  private readonly breakers: Map<string, CircuitBreaker> = new Map();
  private readonly defaultConfig: Partial<Omit<CircuitBreakerConfig, 'name'>>;
  private readonly maxSize: number;

  constructor(redis: Redis, defaultConfig?: Partial<Omit<CircuitBreakerConfig, 'name'>>, maxSize: number = 10000) {
    this.redis = redis;
    this.defaultConfig = defaultConfig || {};
    this.maxSize = maxSize;
  }

  /**
   * Get or create a circuit breaker for a given name.
   * Uses LRU eviction: oldest entries removed when maxSize is exceeded.
   */
  get(name: string, config?: Partial<Omit<CircuitBreakerConfig, 'name'>>): CircuitBreaker {
    let breaker = this.breakers.get(name);

    if (breaker) {
      // Move to end (most recently used) by re-inserting
      this.breakers.delete(name);
      this.breakers.set(name, breaker);
      return breaker;
    }

    // Evict oldest entries if at capacity
    while (this.breakers.size >= this.maxSize) {
      const oldestKey = this.breakers.keys().next().value;
      if (oldestKey !== undefined) {
        this.breakers.delete(oldestKey);
      } else {
        break;
      }
    }

    breaker = new CircuitBreaker(this.redis, {
      name,
      ...this.defaultConfig,
      ...config,
    });
    this.breakers.set(name, breaker);

    return breaker;
  }

  /**
   * Get all circuit breaker statistics
   */
  async getAllStats(): Promise<Map<string, CircuitStats & { name: string }>> {
    const stats = new Map<string, CircuitStats & { name: string }>();

    for (const [name, breaker] of this.breakers) {
      const breakerStats = await breaker.getStats();
      stats.set(name, { ...breakerStats, name });
    }

    return stats;
  }
}

File: services/mail-server/crates/worker-processors/src/common/circuit_breaker.rs
//! Circuit breaker implementation for external service resilience.

use std::sync::atomic::{AtomicU64, Ordering};
use std::sync::RwLock;
use std::time::{Duration, Instant};

/// Circuit breaker state.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum CircuitState {
    /// Circuit is closed, requests flow normally.
    Closed,
    /// Circuit is open, requests are rejected.
    Open,
    /// Circuit is half-open, testing if the service has recovered.
    HalfOpen,
}

/// Circuit breaker configuration.
#[derive(Debug, Clone)]
pub struct CircuitBreakerConfig {
    /// Number of failures before opening the circuit.
    pub failure_threshold: u32,
    /// Duration to keep the circuit open before transitioning to half-open.
    pub open_duration: Duration,
    /// Number of successes in half-open state before closing the circuit.
    pub success_threshold: u32,
    /// Rolling window for failure counting.
    pub window_duration: Duration,
}

impl Default for CircuitBreakerConfig {
    fn default() -> Self {
        Self {
            failure_threshold: 5,
            open_duration: Duration::from_secs(30),
            success_threshold: 3,
            window_duration: Duration::from_secs(60),
        }
    }
}

/// Circuit breaker for protecting against cascading failures.
pub struct CircuitBreaker {
    config: CircuitBreakerConfig,
    state: RwLock<CircuitState>,
    failure_count: AtomicU64,
    success_count: AtomicU64,
    last_failure_time: RwLock<Option<Instant>>,
    opened_at: RwLock<Option<Instant>>,
}

impl CircuitBreaker {
    /// Create a new circuit breaker with the given configuration.
    pub fn new(config: CircuitBreakerConfig) -> Self {
        Self {
            config,
            state: RwLock::new(CircuitState::Closed),
            failure_count: AtomicU64::new(0),
            success_count: AtomicU64::new(0),
            last_failure_time: RwLock::new(None),
            opened_at: RwLock::new(None),
        }
    }

    /// Check if a request is allowed through the circuit.
    pub fn is_allowed(&self) -> bool {
        let mut state = self.state.write().unwrap_or_else(|e| e.into_inner());
        match *state {
            CircuitState::Closed => true,
            CircuitState::Open => {
                // Check if we should transition to half-open
                let opened_at = self.opened_at.read().unwrap_or_else(|e| e.into_inner());
                if let Some(opened) = *opened_at {
                    if opened.elapsed() >= self.config.open_duration {
                        *state = CircuitState::HalfOpen;
                        self.success_count.store(0, Ordering::SeqCst);
                        return true;
                    }
                }
                false
            }
            CircuitState::HalfOpen => true,
        }
    }

    /// Record a successful operation.
    pub fn record_success(&self) {
        let mut state = self.state.write().unwrap_or_else(|e| e.into_inner());
        match *state {
            CircuitState::Closed => {
                // Reset failure count on success
                self.failure_count.store(0, Ordering::SeqCst);
            }
            CircuitState::HalfOpen => {
                let count = self.success_count.fetch_add(1, Ordering::SeqCst) + 1;
                if count >= self.config.success_threshold as u64 {
                    *state = CircuitState::Closed;
                    self.failure_count.store(0, Ordering::SeqCst);
                    self.success_count.store(0, Ordering::SeqCst);
                    let mut opened_at = self.opened_at.write().unwrap_or_else(|e| e.into_inner());
                    *opened_at = None;
                }
            }
            CircuitState::Open => {
                // Shouldn't happen, but reset if it does
            }
        }
    }

    /// Record a failed operation.
    pub fn record_failure(&self) {
        let mut state = self.state.write().unwrap_or_else(|e| e.into_inner());
        match *state {
            CircuitState::Closed => {
                let now = Instant::now();

                // Lock last_failure_time for atomic check-and-update
                let mut last_failure = self
                    .last_failure_time
                    .write()
                    .unwrap_or_else(|e| e.into_inner());
                let should_reset = match *last_failure {
                    Some(last) => now.duration_since(last) > self.config.window_duration,
                    None => true,
                };

                let count = if should_reset {
                    // Reset window - store 1 atomically
                    self.failure_count.store(1, Ordering::SeqCst);
                    1
                } else {
                    self.failure_count.fetch_add(1, Ordering::SeqCst) + 1
                };

                if count >= self.config.failure_threshold as u64 {
                    *state = CircuitState::Open;
                    let mut opened_at = self.opened_at.write().unwrap_or_else(|e| e.into_inner());
                    *opened_at = Some(now);
                }

                *last_failure = Some(now);
            }
            CircuitState::HalfOpen => {
                // Any failure in half-open reopens the circuit
                *state = CircuitState::Open;
                let mut opened_at = self.opened_at.write().unwrap_or_else(|e| e.into_inner());
                *opened_at = Some(Instant::now());
                self.success_count.store(0, Ordering::SeqCst);
            }
            CircuitState::Open => {
                // Already open, update timestamp
                let mut opened_at = self.opened_at.write().unwrap_or_else(|e| e.into_inner());
                *opened_at = Some(Instant::now());
            }
        }
    }

    /// Get the current circuit state.
    pub fn state(&self) -> CircuitState {
        *self.state.read().unwrap_or_else(|e| e.into_inner())
    }

    /// Reset the circuit breaker to closed state.
    pub fn reset(&self) {
        let mut state = self.state.write().unwrap_or_else(|e| e.into_inner());
        *state = CircuitState::Closed;
        self.failure_count.store(0, Ordering::SeqCst);
        self.success_count.store(0, Ordering::SeqCst);
        let mut last_failure = self
            .last_failure_time
            .write()
            .unwrap_or_else(|e| e.into_inner());
        *last_failure = None;
        let mut opened_at = self.opened_at.write().unwrap_or_else(|e| e.into_inner());
        *opened_at = None;
    }
}

impl Default for CircuitBreaker {
    fn default() -> Self {
        Self::new(CircuitBreakerConfig::default())
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_circuit_opens_after_threshold() {
        let cb = CircuitBreaker::new(CircuitBreakerConfig {
            failure_threshold: 3,
            ..Default::default()
        });

        assert!(cb.is_allowed());
        assert_eq!(cb.state(), CircuitState::Closed);

        cb.record_failure();
        cb.record_failure();
        assert!(cb.is_allowed());

        cb.record_failure();
        assert_eq!(cb.state(), CircuitState::Open);
        assert!(!cb.is_allowed());
    }

    #[test]
    fn test_circuit_closes_after_successes() {
        let cb = CircuitBreaker::new(CircuitBreakerConfig {
            failure_threshold: 1,
            success_threshold: 2,
            open_duration: Duration::from_millis(10),
            ..Default::default()
        });

        cb.record_failure();
        assert_eq!(cb.state(), CircuitState::Open);

        // Wait for open duration
        std::thread::sleep(Duration::from_millis(15));

        assert!(cb.is_allowed());
        assert_eq!(cb.state(), CircuitState::HalfOpen);

        cb.record_success();
        cb.record_success();
        assert_eq!(cb.state(), CircuitState::Closed);
    }
}
